File: typescript/hct/hct_test.ts
/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'jasmine';

import * as colorUtils from '../utils/color_utils';

import {Cam16} from './cam16';
import {Hct} from './hct';
import {ViewingConditions} from './viewing_conditions';

const RED = 0xffff0000;
const GREEN = 0xff00ff00;
const BLUE = 0xff0000ff;
const WHITE = 0xffffffff;
const BLACK = 0xff000000;

describe('CAM to ARGB', () => {
  it('red', () => {
    const cam = Cam16.fromInt(RED);

    expect(cam.hue).toBeCloseTo(27.408, 3);
    expect(cam.chroma).toBeCloseTo(113.358, 3);
    expect(cam.j).toBeCloseTo(46.445, 3);
    expect(cam.m).toBeCloseTo(89.494, 3);
    expect(cam.s).toBeCloseTo(91.890, 3);
    expect(cam.q).toBeCloseTo(105.989, 3);
  });

  it('green', () => {
    const cam = Cam16.fromInt(GREEN);

    expect(cam.hue).toBeCloseTo(142.140, 3);
    expect(cam.chroma).toBeCloseTo(108.410, 3);
    expect(cam.j).toBeCloseTo(79.332, 3);
    expect(cam.m).toBeCloseTo(85.588, 3);
    expect(cam.s).toBeCloseTo(78.605, 3);
    expect(cam.q).toBeCloseTo(138.520, 3);
  });

  it('blue', () => {
    const cam = Cam16.fromInt(BLUE);

    expect(cam.hue).toBeCloseTo(282.788, 3);
    expect(cam.chroma).toBeCloseTo(87.231, 3);
    expect(cam.j).toBeCloseTo(25.466, 3);
    expect(cam.m).toBeCloseTo(68.867, 3);
    expect(cam.s).toBeCloseTo(93.675, 3);
    expect(cam.q).toBeCloseTo(78.481, 3);
  });

  it('white', () => {
    const cam = Cam16.fromInt(WHITE);

    expect(cam.hue).toBeCloseTo(209.492, 3);
    expect(cam.chroma).toBeCloseTo(2.869, 3);
    expect(cam.j).toBeCloseTo(100.0, 3);
    expect(cam.m).toBeCloseTo(2.265, 3);
    expect(cam.s).toBeCloseTo(12.068, 3);
    expect(cam.q).toBeCloseTo(155.521, 3);
  });

  it('black', () => {
    const cam = Cam16.fromInt(BLACK);

    expect(cam.hue).toBeCloseTo(0.0, 3);
    expect(cam.chroma).toBeCloseTo(0.0, 3);
    expect(cam.j).toBeCloseTo(0.0, 3);
    expect(cam.m).toBeCloseTo(0.0, 3);
    expect(cam.s).toBeCloseTo(0.0, 3);
    expect(cam.q).toBeCloseTo(0.0, 3);
  });
});

describe('CAM to ARGB to CAM', () => {
  it('red', () => {
    const cam = Cam16.fromInt(RED);
    const argb = cam.toInt();
    expect(argb).toEqual(RED);
  });

  it('green', () => {
    const cam = Cam16.fromInt(GREEN);
    const argb = cam.toInt();
    expect(argb).toEqual(GREEN);
  });

  it('blue', () => {
    const cam = Cam16.fromInt(BLUE);
    const argb = cam.toInt();
    expect(argb).toEqual(BLUE);
  });
});

describe('ARGB to HCT', () => {
  it('green', () => {
    const hct = Hct.fromInt(GREEN);
    expect(hct.hue).toBeCloseTo(142.139, 2);
    expect(hct.chroma).toBeCloseTo(108.410, 2);
    expect(hct.tone).toBeCloseTo(87.737, 2);
  });

  it('blue', () => {
    const hct = Hct.fromInt(BLUE);
    expect(hct.hue).toBeCloseTo(282.788, 2);
    expect(hct.chroma).toBeCloseTo(87.230, 2);
    expect(hct.tone).toBeCloseTo(32.302, 2);
  });

  it('blue tone 90', () => {
    const hct = Hct.from(282.788, 87.230, 90.0);
    expect(hct.hue).toBeCloseTo(282.239, 2);
    expect(hct.chroma).toBeCloseTo(19.144, 2);
    expect(hct.tone).toBeCloseTo(90.035, 2);
  });
});


describe('viewing conditions', () => {
  it('default', () => {
    const vc = ViewingConditions.DEFAULT;
    expect(vc.n).toBeCloseTo(0.184, 3);
    expect(vc.aw).toBeCloseTo(29.981, 3);
    expect(vc.nbb).toBeCloseTo(1.017, 3);
    expect(vc.ncb).toBeCloseTo(1.017, 3);
    expect(vc.c).toBeCloseTo(0.69, 3);
    expect(vc.nc).toBeCloseTo(1.0, 3);
    expect(vc.rgbD[0]).toBeCloseTo(1.021, 3);
    expect(vc.rgbD[1]).toBeCloseTo(0.986, 3);
    expect(vc.rgbD[2]).toBeCloseTo(0.934, 3);
    expect(vc.fl).toBeCloseTo(0.388, 3);
    expect(vc.fLRoot).toBeCloseTo(0.789, 3);
    expect(vc.z).toBeCloseTo(1.909, 3);
  });
});

function colorIsOnBoundary(argb: number): boolean {
  return colorUtils.redFromArgb(argb) === 0 ||
      colorUtils.redFromArgb(argb) === 255 ||
      colorUtils.greenFromArgb(argb) === 0 ||
      colorUtils.greenFromArgb(argb) === 255 ||
      colorUtils.blueFromArgb(argb) === 0 ||
      colorUtils.blueFromArgb(argb) === 255;
}

describe('CamSolver', () => {
  it('returns a sufficiently close color', () => {
    for (let hue = 15; hue < 360; hue += 30) {
      for (let chroma = 0; chroma <= 100; chroma += 10) {
        for (let tone = 20; tone <= 80; tone += 10) {
          const hctColor = Hct.from(hue, chroma, tone);

          if (chroma > 0) {
            expect(Math.abs(hctColor.hue - hue)).toBeLessThanOrEqual(4.0);
          }

          expect(hctColor.chroma).toBeGreaterThanOrEqual(0);
          expect(hctColor.chroma).toBeLessThanOrEqual(chroma + 2.5);

          if (hctColor.chroma < chroma - 2.5) {
            expect(colorIsOnBoundary(hctColor.toInt())).toBe(true);
          }

          expect(Math.abs(hctColor.tone - tone)).toBeLessThanOrEqual(0.5);
        }
      }
    }
  });
});

File: dart/test/hct_test.dart
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:material_color_utilities/hct/cam16.dart';
import 'package:material_color_utilities/hct/hct.dart';
import 'package:material_color_utilities/hct/viewing_conditions.dart';
import 'package:material_color_utilities/utils/color_utils.dart';
import 'package:material_color_utilities/utils/string_utils.dart';
import 'package:test/test.dart';

import './utils/color_matcher.dart';

const black = 0xff000000;
const white = 0xffffffff;
const red = 0xffff0000;
const green = 0xff00ff00;
const blue = 0xff0000ff;
const midgray = 0xff777777;

bool _colorIsOnBoundary(int argb) {
  return ColorUtils.redFromArgb(argb) == 0 ||
      ColorUtils.redFromArgb(argb) == 255 ||
      ColorUtils.greenFromArgb(argb) == 0 ||
      ColorUtils.greenFromArgb(argb) == 255 ||
      ColorUtils.blueFromArgb(argb) == 0 ||
      ColorUtils.blueFromArgb(argb) == 255;
}

void main() {
  test('conversions_areReflexive', () {
    final cam = Cam16.fromInt(red);
    final color = cam.viewed(ViewingConditions.standard);
    expect(color, equals(red));
  });

  test('y_midgray', () {
    expect(18.418, closeTo(ColorUtils.yFromLstar(50.0), 0.001));
  });

  test('y_black', () {
    expect(0.0, closeTo(ColorUtils.yFromLstar(0.0), 0.001));
  });

  test('y_white', () {
    expect(100.0, closeTo(ColorUtils.yFromLstar(100.0), 0.001));
  });

  test('cam_red', () {
    final cam = Cam16.fromInt(red);
    expect(46.445, closeTo(cam.j, 0.001));
    expect(113.357, closeTo(cam.chroma, 0.001));
    expect(27.408, closeTo(cam.hue, 0.001));
    expect(89.494, closeTo(cam.m, 0.001));
    expect(91.889, closeTo(cam.s, 0.001));
    expect(105.988, closeTo(cam.q, 0.001));
  });

  test('cam_green', () {
    final cam = Cam16.fromInt(green);
    expect(79.331, closeTo(cam.j, 0.001));
    expect(108.410, closeTo(cam.chroma, 0.001));
    expect(142.139, closeTo(cam.hue, 0.001));
    expect(85.587, closeTo(cam.m, 0.001));
    expect(78.604, closeTo(cam.s, 0.001));
    expect(138.520, closeTo(cam.q, 0.001));
  });

  test('cam_blue', () {
    final cam = Cam16.fromInt(blue);
    expect(25.465, closeTo(cam.j, 0.001));
    expect(87.230, closeTo(cam.chroma, 0.001));
    expect(282.788, closeTo(cam.hue, 0.001));
    expect(68.867, closeTo(cam.m, 0.001));
    expect(93.674, closeTo(cam.s, 0.001));
    expect(78.481, closeTo(cam.q, 0.001));
  });

  test('cam_black', () {
    final cam = Cam16.fromInt(black);
    expect(0.0, closeTo(cam.j, 0.001));
    expect(0.0, closeTo(cam.chroma, 0.001));
    expect(0.0, closeTo(cam.hue, 0.001));
    expect(0.0, closeTo(cam.m, 0.001));
    expect(0.0, closeTo(cam.s, 0.001));
    expect(0.0, closeTo(cam.q, 0.001));
  });

  test('cam_white', () {
    final cam = Cam16.fromInt(white);
    expect(100.0, closeTo(cam.j, 0.001));
    expect(2.869, closeTo(cam.chroma, 0.001));
    expect(209.492, closeTo(cam.hue, 0.001));
    expect(2.265, closeTo(cam.m, 0.001));
    expect(12.068, closeTo(cam.s, 0.001));
    expect(155.521, closeTo(cam.q, 0.001));
  });

  test('gamutMap_red', () {
    final colorToTest = red;
    final cam = Cam16.fromInt(colorToTest);
    final color =
        Hct.from(cam.hue, cam.chroma, ColorUtils.lstarFromArgb(colorToTest))
            .toInt();
    expect(colorToTest, equals(color));
  });

  test('gamutMap_green', () {
    final colorToTest = green;
    final cam = Cam16.fromInt(colorToTest);
    final color =
        Hct.from(cam.hue, cam.chroma, ColorUtils.lstarFromArgb(colorToTest))
            .toInt();
    expect(colorToTest, equals(color));
  });

  test('gamutMap_blue', () {
    final colorToTest = blue;
    final cam = Cam16.fromInt(colorToTest);
    final color =
        Hct.from(cam.hue, cam.chroma, ColorUtils.lstarFromArgb(colorToTest))
            .toInt();
    expect(colorToTest, equals(color));
  });

  test('gamutMap_white', () {
    final colorToTest = white;
    final cam = Cam16.fromInt(colorToTest);
    final color =
        Hct.from(cam.hue, cam.chroma, ColorUtils.lstarFromArgb(colorToTest))
            .toInt();
    expect(colorToTest, equals(color));
  });

  test('gamutMap_midgray', () {
    final colorToTest = green;
    final cam = Cam16.fromInt(colorToTest);
    final color =
        Hct.from(cam.hue, cam.chroma, ColorUtils.lstarFromArgb(colorToTest))
            .toInt();
    expect(colorToTest, equals(color));
  });

  test('HCT returns a sufficiently close color', () {
    for (var hue = 15; hue < 360; hue += 30) {
      for (var chroma = 0; chroma <= 100; chroma += 10) {
        for (var tone = 20; tone <= 80; tone += 10) {
          final hctRequestDescription = 'H$hue C$chroma T$tone';
          final hctColor = Hct.from(
            hue.toDouble(),
            chroma.toDouble(),
            tone.toDouble(),
          );

          if (chroma > 0) {
            expect(
              hctColor.hue,
              closeTo(hue, 4.0),
              reason: 'Hue should be close for $hctRequestDescription',
            );
          }

          expect(
            hctColor.chroma,
            inInclusiveRange(0.0, chroma + 2.5),
            reason: 'Chroma should be close or less for $hctRequestDescription',
          );

          if (hctColor.chroma < chroma - 2.5) {
            expect(
              _colorIsOnBoundary(hctColor.toInt()),
              true,
              reason: 'HCT request for non-sRGB color should return '
                  'a color on the boundary of the sRGB cube '
                  'for $hctRequestDescription, but got '
                  '${StringUtils.hexFromArgb(hctColor.toInt())} instead',
            );
          }

          expect(
            hctColor.tone,
            closeTo(tone, 0.5),
            reason: 'Tone should be close for $hctRequestDescription',
          );
        }
      }
    }
  });

  group('CAM16 to XYZ', () {
    test('without array', () {
      final colorToTest = red;
      final cam = Cam16.fromInt(colorToTest);
      final xyz = cam.xyzInViewingConditions(ViewingConditions.sRgb);
      expect(xyz[0], closeTo(41.23, 0.01));
      expect(xyz[1], closeTo(21.26, 0.01));
      expect(xyz[2], closeTo(1.93, 0.01));
    });

    test('with array', () {
      final colorToTest = red;
      final cam = Cam16.fromInt(colorToTest);
      final xyz =
          cam.xyzInViewingConditions(ViewingConditions.sRgb, array: [0, 0, 0]);
      expect(xyz[0], closeTo(41.23, 0.01));
      expect(xyz[1], closeTo(21.26, 0.01));
      expect(xyz[2], closeTo(1.93, 0.01));
    });
  });

  group('Color Relativity', () {
    test('red in black', () {
      final colorToTest = red;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(ViewingConditions.make(backgroundLstar: 0.0))
              .toInt(),
          isColor(0xff9F5C51));
    });

    test('red in white', () {
      final colorToTest = red;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(
                  ViewingConditions.make(backgroundLstar: 100.0))
              .toInt(),
          isColor(0xffFF5D48));
    });

    test('green in black', () {
      final colorToTest = green;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(ViewingConditions.make(backgroundLstar: 0.0))
              .toInt(),
          isColor(0xffACD69D));
    });

    test('green in white', () {
      final colorToTest = green;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(
                  ViewingConditions.make(backgroundLstar: 100.0))
              .toInt(),
          isColor(0xff8EFF77));
    });

    test('blue in black', () {
      final colorToTest = blue;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(ViewingConditions.make(backgroundLstar: 0.0))
              .toInt(),
          isColor(0xff343654));
    });

    test('blue in white', () {
      final colorToTest = blue;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(
                  ViewingConditions.make(backgroundLstar: 100.0))
              .toInt(),
          isColor(0xff3F49FF));
    });

    test('white in black', () {
      final colorToTest = white;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(ViewingConditions.make(backgroundLstar: 0.0))
              .toInt(),
          isColor(0xffFFFFFF));
    });

    test('white in white', () {
      final colorToTest = white;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(
                  ViewingConditions.make(backgroundLstar: 100.0))
              .toInt(),
          isColor(0xffFFFFFF));
    });

    test('midgray in black', () {
      final colorToTest = midgray;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(ViewingConditions.make(backgroundLstar: 0.0))
              .toInt(),
          isColor(0xff605F5F));
    });

    test('midgray in white', () {
      final colorToTest = midgray;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(
                  ViewingConditions.make(backgroundLstar: 100.0))
              .toInt(),
          isColor(0xff8E8E8E));
    });

    test('black in black', () {
      final colorToTest = black;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(ViewingConditions.make(backgroundLstar: 0.0))
              .toInt(),
          isColor(0xff000000));
    });

    test('black in white', () {
      final colorToTest = black;
      final hct = Hct.fromInt(colorToTest);
      expect(
          hct
              .inViewingConditions(
                  ViewingConditions.make(backgroundLstar: 100.0))
              .toInt(),
          isColor(0xff000000));
    });
  });
}
